src/kubelet/metric/metric.go

```diff
@@ -18,29 +18,29 @@ import (
 const StatsSummaryPath = "/stats/summary"
 
 // GetMetricsData calls kubelet /stats/summary endpoint and returns unmarshalled response
-func GetMetricsData(c client.HTTPClient) (v1.Summary, error) {
+func GetMetricsData(c client.HTTPClient) (*v1.Summary, error) {
 	resp, err := c.Do(http.MethodGet, StatsSummaryPath)
 	if err != nil {
-		return v1.Summary{}, err
+		return nil, fmt.Errorf("performing GET request to kubelet endpoint: %w", err)
 	}
 	defer resp.Body.Close() // nolint: errcheck
-	if resp.StatusCode != http.StatusOK {
-		return v1.Summary{}, fmt.Errorf("error calling kubelet endpoint. Got status code: %d", resp.StatusCode)
-	}
 
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return v1.Summary{}, fmt.Errorf("error reading the response body of kubelet endpoint. Got error: %v", err.Error())
+		return nil, fmt.Errorf("reading the response body of kubelet endpoint (response code %d): %w", resp.StatusCode, err)
 	}
 
-	var summary = new(v1.Summary)
-	err = json.Unmarshal(body, summary)
-	if err != nil {
-		return v1.Summary{}, fmt.Errorf("error unmarshaling the response body. Got error: %v", err.Error())
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("received non-OK response code from kubelet: %d. Response body: %v", resp.StatusCode, body)
 	}
 
-	return *summary, nil
+	summary := &v1.Summary{}
 
+	if err := json.Unmarshal(body, summary); err != nil {
```

Comment:
Could we consider using `json.NewDecoder(resp.Body).Decode(summary)`? I think it's easier to read if we are not doing weird stuff with the body.

Reply:
> It might additionally be easier on memory for large payloads.
This comment was intended to go here.

Reply:
We could, though then we lose the ability of reading the error message from the request, which might be for example:
```
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* Connection state changed (MAX_CONCURRENT_STREAMS == 250)!
< HTTP/2 403
< content-type: text/plain; charset=utf-8
< content-length: 92
< date: Fri, 16 Jul 2021 06:15:09 GMT
<
* Connection #0 to host 10.0.0.2 left intact
Forbidden (user=system:kube-controller-manager, verb=get, resource=nodes, subresource=stats)
```

Reply:
We could check the status code, and if it is non-200 then print the body along the error. It seems reasonable to assume that a 200 status will come with a valid JSON payload.

Reply:
That creates more cases, as error handling will have to read the body in a different way than the regular path. It seems more complicated to me without clear benefit.